### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by Megha-Joshi**
Screenshot (158)
Screenshot (159)

This PR includes-
- Login Authentication
- Logout Authentication
- Signin Authentication
- Public and Private Routes added
- Login, signup, like, history, watchLater pages added

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -2,12 +2,34 @@ import "./App.css";
 import logo from "./logo.png";
 import { Homepage } from "./pages/Homepage/homepage";
 import { Landingpage } from "./pages/Landingpage/landingpage";
+import { Login } from "./pages/Login-Signup/login";
+import { Signup } from "./pages/Login-Signup/signup";
+import { Route, Routes } from "react-router-dom";
+import { Like } from "./pages/Liked/like";
+import { History } from "./pages/History/history";
+import { WatchLater } from "./pages/WatchLater/watchLater";
```
Comment (dev-enforced) [inline src/App.js:10]: Can use barrel exports to develop the component related to importing multiple pages in a single line by making an index.js file in pages folder that exports those components from there itself.

File: src/context/authContext.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    const loginHandler = async ( user ) =>{
+        console.log("entered login")
+        try{
+            const response = await loginAPI(user)
+            if (response.status === 200){
+                localStorage.setItem("token", response.data.encodedToken);
+                localStorage.setItem("user", JSON.stringify(response.data.foundUser));
+                setToken(response.data.encodedToken);
+                setUser(response.data.foundUser);
+                console.log("in login handler")
```
Comment (dev-enforced) [inline src/context/authContext.js:24]: Can remove this debugger

File: src/context/authContext.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    //             localStorage.setItem("user", JSON.stringify(response.data.createdUser));
+    //             setToken(response.data.encodedToken);
+    //             setUser(response.data.createdUser);
+
+    //             navigate("/homepage")
+    //         }
+    //     }catch(error){
+    //         console.log(error);
+    //     }
+    // }
```
Comment (dev-enforced) [inline src/context/authContext.js:65]: Can remove this commented block if it is not required further

File: src/context/authContext.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+import { useNavigate } from "react-router-dom";
+import {loginAPI, signupAPI} from "../services/authAPI.js";
+import "../pages/Login-Signup/login.jsx";
+import "../pages/Login-Signup/signup.jsx";
+
+const AuthContext = createContext();
+
+const AuthProvider = ({children}) => {
+    const navigate = useNavigate();
+    const x = JSON.parse(localStorage.getItem("user"))
```
Comment (dev-enforced) [inline src/context/authContext.js:11]: Can rename the variable to `const userDetails` for better naming

File: src/video-components/privateRoute/privateRoute.jsx [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import {Navigate,useLocation,Outlet} from "react-router-dom" 
+import { useAuth } from "../../context/authContext"
+
+export const PrivateRoute = () => {
+   const {token} = useAuth();
+   const location = useLocation();
+
+   return token ? (
+       <Outlet/>
```
Comment (dev-enforced) [inline src/video-components/privateRoute/privateRoute.jsx:9]: This is a new approach I got to know about Megha!! Good work on using library provided options for redirection tasks!!

File: src/pages/Login-Signup/login.jsx [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+const navigate = useNavigate();
+const [ userDetails, setUserDetails] = useState({email: "", password: ""});
+
+// const guestUserDetails = (e) =>{
+//     e.preventDefault();
+// setUserDetails({
+// email: "adarshbalika@gmail.com",
+// password: "adarshBalika123"
+// })
+// }
```
Comment (dev-enforced) [inline src/pages/Login-Signup/login.jsx:18]: Can remove this block of code!!

File: src/pages/Login-Signup/login.jsx [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+function formHandler(e){
+    e.preventDefault();
+      loginHandler(userDetails);
+}
+
+// useEffect(()=>{
+// if(token){
+// navigate("/homepage");
+// }
+// },[token,navigate]);
```
Comment (dev-enforced) [inline src/pages/Login-Signup/login.jsx:37]: Similarly can remove this block too if not required. I guess you are trying to implement this feature right related to authenticated users trying to access login and signup routes then they shall be navigated somewhere else right?

File: src/pages/WatchLater/watchLater.jsx [hunk 8]
```diff
@@ -0,0 +1,61 @@
+import { useState } from "react";
+import { Navbar } from "../../video-components/Navbar/navbar";
+import { Sidebar } from "../../video-components/Sidebar/sidebar";
```
Comment (dev-enforced) [inline src/pages/WatchLater/watchLater.jsx:3]: Can implement barrel exports here too

File: src/pages/Login-Signup/login.jsx [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+return(
+<div className="login-container justify-align">
+    <form className="form-container" onSubmit={formHandler}>
+        <h2 className="login-head">Login</h2>
+        <fieldset>
+            <legend for="username" className="inp-txt">Email</legend>
+            <input type="text" id="username" name="username" placeholder="abc@gmail.com" className="input-box"
+                value={userDetails.email} onChange={(e) => setUserDetails({...userDetails, email: e.target.value})} required /><br />
```
Comment (dev-enforced) [inline src/pages/Login-Signup/login.jsx:46]: Try avoid using `<br/>` tags while developing components!!
One more thing is that the onChange for both the fields have similar functions itself so you can write a common function itself and pass it in the onChange itself that just takes the event itself and updates userDetails by taking NAME and target of that event.
You can ask queries if you have difficulties framing it up

**Review by dev-enforced — APPROVED** [review]
The UI of authentication pages looks really good. Got introduced to this new support from React Router Library related to Outlet for redirecting to protected / private sections.
Looking forward to see the live preview of Goosebumps soon . Would help the reviewers give better feedback on it!! Code suggestions are added below:

File: src/App.js (hunk 1 above)
Comment (Megha-Joshi) [reply src/App.js:10]: Okay, will do it!

File: src/context/authContext.js (hunk 2 above)
Comment (Megha-Joshi) [reply src/context/authContext.js:24]: Okay

File: src/context/authContext.js [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    const [ token, setToken ] = useState(localStorage.getItem("token"));
+    const [user, setUser] = useState(x);
+
+    const loginHandler = async ( user ) =>{
+        console.log("entered login")
```
Comment (zoyatanweer) [inline src/context/authContext.js:16]: remove console.log

File: src/context/authContext.js [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+            }
+        }
+        catch(err){
+            console.log(err)
+        }
+
+    }
+
+    const signupHandler = async (user) => {
+        console.log("entered signup")
```
Comment (zoyatanweer) [inline src/context/authContext.js:35]: remove console.log

File: src/pages/Login-Signup/auth.css [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+.text-btn{
+    background-color: transparent;
+    border: none;
+    color: var(--primary-color);
+    font-size: 1rem;
+    cursor: pointer;
+}
+
+.login-btn{
+    background-color: #818A91;
```
Comment (zoyatanweer) [inline src/pages/Login-Signup/auth.css:78]: use var for color

File: src/pages/Login-Signup/login.jsx (hunk 7 above)
Comment (Megha-Joshi) [reply src/pages/Login-Signup/login.jsx:37]: Yes, and then forget to remove it.

File: src/pages/WatchLater/watchLater.jsx (hunk 8 above)
Comment (Megha-Joshi) [reply src/pages/WatchLater/watchLater.jsx:3]: Okay

File: src/pages/Login-Signup/login.jsx (hunk 9 above)
Comment (Megha-Joshi) [reply src/pages/Login-Signup/login.jsx:46]: Okay

**Review by zoyatanweer — APPROVED** [review]
- code looks clean and very nice use of comments. It made easier to understand
- make the suggested changes and your code is then good to go
- good to merge

**Pull request merged**